Automattic/wp-calypso
# Checkout: unify logic that determines if web payment methods are available

#### Problem

When checkout determines which payment methods to display, it uses the custom React hook `useIsWebPayAvailable()` for "web payment" methods (Apple Pay and Google Pay). That hook creates a Stripe Payment Request so we can ask Stripe if Apple and/or Google Pay should be allowed. If it says they are, then we create the payment method UI objects for those payment methods and list them as available. When a user then selects a web payment method, that payment method UI will call the hook `useStripePaymentRequest()` to create another Stripe Payment request that allows it to show the special submit button for that payment type.

The Stripe Payment Request created by `useIsWebPayAvailable()` is slightly different from the one created by `useStripePaymentRequest()`; the former does not include `displayItems`, for example.

Sometimes, the call to `useIsWebPayAvailable()` says that a payment method is available, but then `useStripePaymentRequest()` says that it is not available once the user has clicked to select the payment method. This results in us failing to render the submit button and causing the user to see an error message, which is very poor UX. It's not clear why this could happen since the logic that determines what payment methods are available is obfuscated behind Stripe and Apple/Google's servers, but the frequency of this error has been increasing lately (see https://github.com/Automattic/wp-calypso/issues/70019).

#### Proposed Changes

In this PR, we remove `useIsWebPayAvailable()` and instead use only `useStripePaymentRequest()` for both determining which payment methods are available as well as displaying the buttons for those payment methods. Hopefully this will decrease the instances where users are getting an error.

#### Testing Instructions

To test Google Pay:

- Add a product to your cart and visit checkout _in production calypso_.
- Make sure you see Google Pay in checkout in production. If not, you may need to add a card to your Google wallet.
- Use the calypso.live link in this diff to load calypso; this is necessary so that you can view the app over https.
- Visit checkout again in the calypso.live version of calypso.
- Verify that you see Google Pay as a payment method.
